File: diploma-rest-api/src/main/java/org/arhor/diploma/web/model/message/Message.java
package org.arhor.diploma.web.model.message;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import lombok.Data;

import java.util.List;

@Data
@JsonPropertyOrder({"code", "severity", "text", "details"})
public final class Message {

  private final int code;
  private final Severity severity;
  private final String text;
  private final List<String> details;

  public static Builder info() {
    return new Builder(Severity.INFO);
  }

  public static Builder warn() {
    return new Builder(Severity.WARN);
  }

  public static Builder error() {
    return new Builder(Severity.ERROR);
  }

  private static final class Builder {

    private final Severity severity;

    private int code;
    private String text;
    private List<String> details;

    private Builder(Severity severity) {
      this.severity = severity;
    }

    public Message build() {
      return new Message(code, severity, text, details);
    }

    public Builder withCode(int code) {
      this.code = code;
      return this;
    }

    public Builder withText(String text) {
      this.text = text;
      return this;
    }

    public Builder withDetails(String... details) {
      this.details = List.of(details);
      return this;
    }
  }
}

File: diploma-server/server-app/src/main/kotlin/org/arhor/diploma/web/model/Message.kt
package org.arhor.diploma.web.model

import com.fasterxml.jackson.annotation.JsonPropertyOrder
import java.time.LocalDateTime

@JsonPropertyOrder("severity", "timestamp", "code", "text", "details")
data class Message(
    val severity: Severity,
    val timestamp: LocalDateTime,
    val code: ErrorCode,
    val text: String? = null,
    val details: List<Any>?
) {

    enum class Severity { INFO, WARN, ERROR }
}

/**
 * Represents general message response from the server.
 *
 * JSON form:
 * ```
 * {
 *   "messages": [
 *     {
 *       "severity": "ERROR",
 *       "timestamp": "2021-02-24T13:03:24.7504",
 *       "code": "GEN-00000",
 *       "text": "Validation failed",
 *       "details": [
 *         "field 'username' should not be blank",
 *         "field 'password' should not be null"
 *       ]
 *     }
 *   ]
 * }
 * ```
 */
data class MessageResponse(val messages: List<Message>)

class MessageBuilder(
    val severity: Message.Severity,
    var timestamp: LocalDateTime = LocalDateTime.now(),
    var code: ErrorCode = ErrorCode.UNCATEGORIZED,
    var text: String? = null,
    var details: List<Any?> = emptyList()
)

@DslMarker
annotation class MessageResponseMaker

@MessageResponseMaker
class MessageResponseBuilder {

    private val messages = mutableListOf<Message>()

    fun info(init: MessageBuilder.() -> Unit) = initMessage(Message.Severity.INFO, init)

    fun warn(init: MessageBuilder.() -> Unit) = initMessage(Message.Severity.WARN, init)

    fun error(init: MessageBuilder.() -> Unit) = initMessage(Message.Severity.ERROR, init)

    fun build(): MessageResponse {
        return MessageResponse(messages.toList())
    }

    private inline fun initMessage(severity: Message.Severity, init: MessageBuilder.() -> Unit) {
        val builder = MessageBuilder(severity)

        builder.init()

        messages.add(
            Message(
                severity = builder.severity,
                timestamp = builder.timestamp,
                code = builder.code,
                text = builder.text,
                details = builder.details.filterNotNull()
            )
        )
    }
}

inline fun messageResponse(init: MessageResponseBuilder.() -> Unit): MessageResponse {
    val message = MessageResponseBuilder()
    message.init()
    return message.build()
}
